File: src/typescript/utils/account.ts
import algosdk from "algosdk";
import nacl from "tweetnacl";
import { masterToDerivation } from "./wallet";
import hkdf from "futoin-hkdf";
import sha512 from "js-sha512";
import * as utils from "./encoders";
import base32 from "hi-base32";

export const generateAccountList = (
  seed: string,
  index: number,
  amount: number
): string[] => {
  let accountArray: string[] = [];
  let decodeKey: Uint8Array;

  if (seed.length > 50) {
    let masterKey = masterToDerivation(seed);
    decodeKey = masterKey;
  } else {
    decodeKey = new Uint8Array(Buffer.from(seed, "base64"));
  }

  for (var i = index; i < index + amount; i++) {
    let info = "AlgorandDeterministicKey-1";

    let key = hkdf.expand(
      "SHA512-256",
      hkdf.hash_length("sha256"),
      Buffer.from(decodeKey),
      32,
      info
    );

    let { publicKey } = nacl.sign.keyPair.fromSeed(key);

    const checksum = sha512.sha512_256.array(publicKey).slice(28, 32);

    const addr = base32.encode(utils.concatArrays(publicKey, checksum));

    accountArray.push(addr.toString().slice(0, 58));
  }

  return accountArray;
};

export const generateAccountMnemonic = (
  seed: string,
  index: number
): string[] => {
  let accountArray: string[] = [];
  let decodeKey: Uint8Array;

  if (seed.length > 50) {
    let masterKey = masterToDerivation(seed);
    decodeKey = masterKey;
  } else {
    decodeKey = new Uint8Array(Buffer.from(seed, "base64"));
  }

  for (var i = index; i < index + 1; i++) {
    let info = "AlgorandDeterministicKey-1";

    let key = hkdf.expand(
      "SHA512-256",
      hkdf.hash_length("sha256"),
      Buffer.from(decodeKey),
      32,
      info
    );

    let { secretKey } = nacl.sign.keyPair.fromSeed(key);

    accountArray.push(algosdk.secretKeyToMnemonic(secretKey));
  }

  return accountArray;
};

File: src/go/utils/account.go
package utils

import (
	"fmt"
	b64 "encoding/base64"
	"encoding/base32"
	"crypto/sha512"
	"crypto/ed25519"
	"golang.org/x/crypto/hkdf"
	"github.com/algorand/go-algorand-sdk/mnemonic"
)


func GenerateAccountList(seed string, index int, amount int)([]string){
	var accountArray []string;
	var decodedKey []byte;
	var err error;

	if amount == 0 {
		amount = 1
	}

	if len(seed) > 50 {
		masterKey := MasterToDerivation(seed)
		fmt.Println(masterKey)
		
		decodedKey, err = b64.StdEncoding.DecodeString(masterKey)
		
		if err != nil {
			fmt.Println(err )
		}
	} else {  
		decodedKey, err = b64.StdEncoding.DecodeString(seed)
		if err != nil {
			fmt.Println(err )
		}
	}
	
	for i := index; i < index + amount; i++ {
		info := []byte(fmt.Sprintf("AlgorandDeterministicKey-%d", i))

		hash := sha512.New512_256

		keystream := hkdf.Expand(hash, decodedKey, info)

		pub, priv, err := ed25519.GenerateKey(keystream)

		if err != nil{
			// Just to ignore the priv warning
			fmt.Println(priv)
			fmt.Println(err)
		}

		chksum := sha512.Sum512_256(pub[:])

		checksumAddress := append(pub[:], chksum[28:]...)
		
		accountArray = append(accountArray, base32.StdEncoding.EncodeToString(checksumAddress)[:58])
	}

	return accountArray
}


func GenerateAccountMnemonic(seed string, index int)([]string){
	var accountArray []string;
	var decodedKey []byte;
	var err error;

	if len(seed) > 50 {
		masterKey := MasterToDerivation(seed)

		decodedKey, err = b64.StdEncoding.DecodeString(masterKey)
		
		if err != nil {
			fmt.Println(err )
		}
	} else {  
		decodedKey, err = b64.StdEncoding.DecodeString(seed)
		if err != nil {
			fmt.Println(err )
		}
	}
	
	for i := index; i < index + 1; i++ {
		info := []byte(fmt.Sprintf("AlgorandDeterministicKey-%d", i))

		hash := sha512.New512_256

		keystream := hkdf.Expand(hash, decodedKey, info)

		pub, priv, err := ed25519.GenerateKey(keystream)

		if err != nil{
			// Just to ignore the priv warning
			fmt.Println(pub)
			fmt.Println(err)
		}

		mnemonic, err := mnemonic.FromPrivateKey(priv)
		accountArray = append(accountArray, mnemonic)
	}

	return accountArray
}